src/test/java/com/linecorp/armeria/client/manipulating/HttpHeaderManipulatingClientBuilderTest.java

```diff
@@ -0,0 +1,224 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.manipulating;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+
+import java.util.List;
+
+import org.junit.Test;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.http.DefaultHttpRequest;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpMethod;
+import com.linecorp.armeria.common.http.HttpRequest;
+
+import io.netty.util.AsciiString;
+
+
+public class HttpHeaderManipulatingClientBuilderTest {
+
+    @Test
+    public void testErrorOnNullOrEmptyHeader() {
+        HttpHeaderManipulatingClientBuilder builder = newBuilder();
+        boolean thrown = false;
+        try {
+            builder.add(null, "value");
+            fail("Exception not thrown on null header ");
+
+            builder.add(EMPTY_HEADER, "value");
+            fail("Exception not thrown on empty header");
+
+            builder.set(null, "value");
+            fail("Exception not thrown on null header");
+
+            builder.set(EMPTY_HEADER, "value");
+            fail("Exception not thrown on empty header");
+
+            builder.remove((AsciiString) null);
+            fail("Exception not thrown on null header");
+
+            builder.remove(EMPTY_HEADER);
+            fail("Exception not thrown on empty header");
+
+            builder.update(null);
+            fail("Exception not thrown on null updater");
+
+        } catch (Exception e) {
+            thrown = true;
+        }
+        assertEquals(true, thrown);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static final Client dummyDelegate = mock(Client.class);
+    private static final ClientRequestContext dummyContext = mock(ClientRequestContext.class);
+
+    @Test
+    public void testEmpty() throws Exception {
+        HttpRequest request = newRequest();
+        Client client = newBuilder().buildClient(dummyDelegate);
+        client.execute(dummyContext, request);
+        assertEquals(3, request.headers().size());
+    }
+
+    @Test
+    public void testAdd() throws Exception {
+        HttpRequest request = newRequest();
+        Client client = newBuilder()
+            .add(HEADER_1, "a1")
+            .add(HEADER_0, "a0")
+            .buildClient(dummyDelegate);
+
+        client.execute(dummyContext, request);
+        HttpHeaders headers = request.headers();
+        
+        List<String> value0 = headers.getAll(HEADER_0);
+        List<String> value1 = headers.getAll(HEADER_1);
+
+        assertEquals("[v0, a0]", value0.toString());
```

Comment:
Please compare list directly intead of using `toString()` result. Maybe, AssertJ is helpful to you.